RFC.md

```diff
@@ -1,12 +1,295 @@
+alias analysis can use this information to model multidimensional-array
+semantics.
+
+There is a more realistic (and involved) example in [Appendix A](#Appendix-A)
+in the same spirit as the above simple example, but one a compiler might
+realistically wish to perform.
+
+## Evaluation of the impact of the intrinsic on accuracy of dependence analysis
+
+- This has been implemented in an exprimental branch of Polly, and was used 
+on the COSMO climate weather model. This greatly helped increase the accuracy
+of Polly's analysis, since we eliminated the guessing game from the array analysis.
+
+- This has also been implemented as part of a GSoC effort to unify
+Chapel and Polly. 
+
+- Molly, the distributed version of Polly written by Michael Kruse for his PhD
+  also implemented a similar scheme. In his use-case, optimistic run-time checks
+  with delinearization was not possible, so this kind of intrinsic was _necessary_
+  for the application, not just _good to have_. More details are available
+  in his PhD thesis: [Lattice QCD Optimization and Polytopic Representations of Distributed Memory](https://tel.archives-ouvertes.fr/tel-01078440).
+  In particular, Chapter 9 contains a detailed discussion.
+
+# Representations
+
+## Instruction / Intrinsic
+
+### Syntax
+```
+<result> = multidim_array_index <ty> <ty>* <ptrval> {<stride>, <idx>}*
+```
+
+### Overview:
+The `multidim_array_index` instruction is used to get the address of 
+an element from an array. It performs address calcuation only and 
+does not access memory. It is similar to `getelementptr`. However, it imposes
+additional semantics which allows the optimiser to provide better optimisations
+than `getlementptr`.
+
+
+
+### Arguments:
+The first argument is always a type used as the basis for the calculations. The
+second argument is always a pointer, and is the base address to start the
+calculation from. The remaining arguments are a list of pairs. Each pair
+contains a dimension stride, and an offset with respect to that stride.
+
+
+### Semantics:
+
+`multidim_array_index` represents a multi-dimensional array index, In particular, this will
+mean that we will assume that all indices `<idx_i>` are non-negative.
+
+Additionally, we assume that, for each `<str_i>`, `<idx_i>` pair, that 
+`0 <= idx_i < str_i`.
+
+Optimizations can assume that, given two multidim_array_index instructions with matching types:
+
+```
+multidim_array_index <ty> <ty>* <ptrvalA> <strA_1>, <idxA_1>, ..., <strA_N>, <idxA_N>
+multidim_array_index <ty> <ty>* <ptrvalB> <strB_1>, <idxB_1>, ..., <strB_N>, <idxb_N>
+```
+
+If `ptrvalA == ptrvalB` and the strides are equal `(strA_1 == strB_1 && ... && strA_N == strB_N)` then:
+
+- If all the indices are equal (that is, `idxA_1 == idxB_1 && ... && idxA_N == idxB_N`), 
+  then the resulting pointer values are equal.
+
+- If any index value is not equal (that is, there exists an `i` such that `idxA_i != idxB_i`),
+  then the resulting pointer values are not equal.
```

Comment:
The `inrange` modifier of GEP has this semantics. Maybe use a similar wording:

0 <= idx_i < str_i